docs/topics/db/queries.txt

```diff
@@ -937,6 +937,42 @@ To query for missing keys, use the ``isnull`` lookup::
     On PostgreSQL, if only one key or index is used, the SQL operator ``->`` is
     used. If multiple operators are used then the ``#>`` operator is used.
 
+Querying optional JSON keys
+--------------------------------------
+
+When querying by an optional JSON key, any records which lack that key will be
+excluded from the results::
+
+    >>> Dog.objects.create(name='Rufus', data={'breed': 'collie', 'owner': 'Bob'})
+    <Dog: Rufus>
+    >>> Dog.objects.create(name='Meg', data={'breed': 'collie'})
+    <Dog: Meg>
+    >>> Dog.objects.create(name='Archie', data={'owner': 'Bob'})
+    <Dog: Archie>
+    >>> Dog.objects.create(name='Max', data={'breed': 'labrador', 'owner': 'Alice'})
+    <Dog: Max>
+    >>> Dog.objects.filter(data__breed='collie')
+    <QuerySet [<Dog: Rufus>, <Dog: Meg>]>
+    >>> Dog.objects.exclude(data__breed='collie')
+    <QuerySet [<Dog: Max>]>
+
+This means that the union between a query and its negation may not return every
+instance of a model::
+
+    >>> Dog.objects.count()
+    4
+    >>> Dog.objects.filter(data__breed='collie').count()
+    2
+    >>> Dog.objects.exclude(data__breed='collie').count()
+    1
+
+A negated filter will still return the same objects as an exclude::
+
+    >>> Dog.objects.filter(~Q(data__owner='Bob'))
```

Comment:
Should we also note that the query that returns a proper negation of `data__owner='Bob'` would be `~Q(data__owner='Bob') | Q(data__owner__isnull=True)`?

Reply:
Absolutely! Fixed